pkg/controller/stats/composite.go

```diff
@@ -0,0 +1,302 @@
+			row = append(row, "", "", "", "", "", "", "")
+		} else {
+			row = append(row, strconv.FormatUint(uint64(stat.RealmStats.CodesIssued), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.RealmStats.CodesClaimed), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.RealmStats.CodesInvalid), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.RealmStats.TokensClaimed), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.RealmStats.TokensInvalid), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.RealmStats.CodeClaimMeanAge.Duration.Seconds()), 10))
+			row = append(row, strings.Join(stat.RealmStats.CodeClaimAgeDistributionAsStrings(), "|"))
+		}
+		if stat.KeyServerStats == nil {
+			// no key sever stats, 8 empty columns
+			row = append(row, "", "", "", "", "", "", "", "")
+		} else {
+			row = append(row, strconv.FormatUint(uint64(stat.KeyServerStats.PublishRequests.UnknownPlatform), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.KeyServerStats.PublishRequests.Android), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.KeyServerStats.PublishRequests.IOS), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.KeyServerStats.TotalTEKsPublished), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.KeyServerStats.RevisionRequests), 10))
+			row = append(row, strconv.FormatUint(uint64(stat.KeyServerStats.RequestsMissingOnsetDate), 10))
+			row = append(row, join(stat.KeyServerStats.TEKAgeDistribution, "|"))
+			row = append(row, join(stat.KeyServerStats.OnsetToUploadDistribution, "|"))
+		}
+
+		if err := w.Write(row); err != nil {
+			return nil, fmt.Errorf("failed to write CSV entry %d: %w", i, err)
+		}
+	}
+
+	w.Flush()
+	if err := w.Error(); err != nil {
+		return nil, fmt.Errorf("failed to create CSV: %w", err)
+	}
+
+	return b.Bytes(), nil
+}
+
+// HandleComposite returns composite states for realm + key server
+// The key server stats may be omitted if that is not enabled
+// on the realm.
+func (c *Controller) HandleComposite(typ Type) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		currentRealm, ok := authorizeFromContext(ctx, rbac.StatsRead)
+		if !ok {
+			controller.Unauthorized(w, r, c.h)
+			return
+		}
+
+		realmStats, err := currentRealm.StatsCached(ctx, c.db, c.cacher)
+		if err != nil {
+			controller.InternalError(w, r, c.h, err)
+			return
+		}
+
+		var stats CompositeStats = make([]*CompositeDay, 0, len(realmStats))
+		statsMap := make(map[time.Time]*CompositeDay, len(realmStats))
+		for _, rs := range realmStats {
+			day := &CompositeDay{
+				Day:        rs.Date,
+				RealmStats: rs,
+			}
+			stats = append(stats, day)
+			statsMap[rs.Date] = day
+		}
+
+		if c.db.HasKeyServerStats(currentRealm.ID) {
+			days, err := c.db.ListKeyServerStatsDaysCached(ctx, currentRealm.ID, c.cacher)
+			if err != nil {
+				controller.InternalError(w, r, c.h, err)
+				return
+			}
+
+			needsSort := false
```

Comment:
This feels like an unnecessary optimization. `sort.Slice` is O(n) for a sorted slice.